test/integration_test/s3/getObject.spec.js

```diff
@@ -0,0 +1,209 @@
+var helpers = require('../helpers');
+var AWS = helpers.AWS;
+var bucketName = helpers.sharedBucket;
+describe('download integrity', function() {
+  var params;
+  var s3;
+  function putObject(customParams, done) {
+    params = {
+      Bucket: customParams.Bucket || params.Bucket,
+      Key: customParams.Key || params.Key,
+    };
+    s3.createBucket({Bucket: params.Bucket}).promise().then(function() {
+      return s3.waitFor('bucketExists', {Bucket: params.Bucket}).promise()
+    }).then(function() {
+      return s3.putObject({
+        Bucket: params.Bucket,
+        Key: params.Key,
+        Body: customParams.Body || params.Body
+      }).promise();
+    }).then(function() {
+      return s3.waitFor('objectExists', {
+        Bucket: params.Bucket,
+        Key: params.Key,
+      }).promise();
+    }).then(function() { 
+      if (typeof done === 'function') done();
+    }).catch(function(err) {
+      throw new Error('Cannot put object: ' + err);
+      exit(1);
+    });
+  }
+
+  function delectBucket(done) {
+    s3.listObjectVersions({
+      Bucket: params.Bucket
+    }).promise().then(function(data) {
+      var removeObjectsParams = {
+        Bucket: params.Bucket,
+        Delete: {
+          Quiet: false,
+          Objects: []
+        }
+      }
+      for (var version of data.Versions) {
+        removeObjectsParams.Delete.Objects.push({
+          Key: version.Key,
+          VersionId: version.VersionId === 'null' ? null : version.VersionId,
+        });
+      }
+      return removeObjectsParams;
+    }).then(function(removeObjectsParams) {
+      return s3.deleteObjects(removeObjectsParams).promise();
+    }).then(function() {
+      return s3.waitFor('objectNotExists', {
+        Bucket: params.Bucket,
+        Key: params.Key,
+      }).promise()
+    }).then(function() {
+      return s3.deleteBucket({Bucket: params.Bucket}).promise();
+    }).then(function(data) {
+      return s3.waitFor('bucketNotExists', {Bucket: params.Bucket}).promise();
+    }).then(function() {
+      if (typeof done === 'function') done();
+    }).catch(function(err) {
+      throw new Error('Cannot delete bucket: ' + err);
+      exit(1);
+    })
+  }
+
+  before('setup s3 client with responseChecksumAlgorithm equals \'md5\'', function() {
+    s3 = new AWS.S3({responseChecksumAlgorithm: 'md5'});
```

Comment:
Why don't you create the bucket used by all the tests in the `before` step? We shouldn't have to create a new bucket for every single test, just this suite of tests. 

I also wouldn't mix the `putObject` method with `createBucket`. `putObject` is doing too much, and adds 'global' (across tests) state. For example, you don't directly pass it the bucket or key, instead relying on a closure that every test has access to (and can change). That could lead to tricky edge cases coming up later that are hard to debug.

Reply:
Yes, you can pass in the bucket and key, but right, it will change the global state. But capsulizing them will also guarantee `putObject` always succeed(bucket always there).

Reply:
But you could also make sure the bucket is there in the `before` hook. Just call `done()` after the `waitFor` method completes. Then you also only need to create it once; I don't think there's a reason we need to create a new bucket for every test since we aren't testing any bucket-specific configurations here. Creating a new bucket with each test also creates more points of failure (rate/resource limits, for example).

Reply:
yea. I will move the `putObject` to `before` trait. And use timestamp in bucket name. I was thinking that we can write less code if we are going to add more tests here. But now I agree that this is an overkill and not worthwhile.